doc/adr/0012-modules-structure.md

```diff
@@ -0,0 +1,104 @@
+- `core` is the Global L0 module (but also contains state channels that must be extracted to be reused in other modules),
+- `dag-l1` is the DAG L1 module.
+
+```mermaid
+graph TD
+    A[Shared]
+    B[Keytool]
+    C[Kernel]
+    D[Wallet]
+    E[SDK]
+    F[DAG Shared]
+    G[Core]
+    H[DAG L1]
+    
+    B --> A
+    C --> A
+    D --> B
+    D --> A
+    E --> C
+    E --> B
+    E --> A
+    E --> F
+    F --> C
+    F --> B
+    F --> A
+    G --> E
+    G --> C
+    G --> B
+    G --> A
+    G --> F
+    H --> F
+    H --> E
+    H --> C
+    H --> B
+    H --> A
+```
+
+Furthermore, dependencies were partially explicitly declared and partially transient. After some time, it became unmaintainable.
+
+The second reason for restructuring modules is the fact that 3rd parties cannot use the SDK module in the current form. It does not expose all required operations, so internal modules were used directly, leaking the project's internal structure and locking further internal changes.
+
+## Decision
+
+To simplify the structure of modules, we will split `shared` into two separate modules - the first containing scala and libraries extensions, the latter containing the tessellation schema. `keytool` module will be splitted into: operations/functions that we will put to `schema` and exposing an app that will be combined into `wallet` module.
+Then, all the things related to node operations would be kept in `node-shared` module (including gossip protocol, consensus, etc.). This leads to emptying the `sdk` module, so we will be able to implement the `sdk` module as a proxy between the external world and the internal tessellation structure. `sdk` module should reexpose all necessary operations to be used by 3rd parties. For example - to create a keypair, global L0 can use the `schema` module directly, but 3rd party will use the `sdk` module that will reexpose keypair creation function. That way, the internal structure of tessellation will be hidden behind the facade.
+
+Modules will be transient - for example: `wallet` does not need to depend on `shared` directly, because it depends on `schema` that depends on `shared`.
+
+A separate ADR will organize dependencies (libraries) of modules.
+
+- `shared` will contain base classes and extensions (like kryo or refined extensions),
+- `schema` will contain tessellation schema (like Address, Balance, GlobalSnapshot, base Block or Transaction, base State Channels types) and key-related operations,
+- `wallet` will contain operations required by the wallet-cli (like creating transactions), but also will expose key operations (like creating a new key)
+- `dag-shared` will contain shared types and classes between Global L0 and DAG L1, related to $DAG (like $DAG block or $DAG transaction),
+- `node-shared` will contain all the operations required to launch a node (like base TessellationIO app, gossip protocol, cluster management or joining algorithm),
+- `core` will remain the Global L0 module,
```

Comment:
Shouldn't we rename it to `global-l0` or something?

Reply:
imo it will make more burden but `core` also is a valid name, as the Global Layer 0 is the most underlying (base) layer, so a core for everything else.

Reply:
Imo `core` is fine because `global-l0` sounds like there was `global-l1` as well

Reply:
For me it's a layer that brings all the layers (those which decide to use global L0 of course) together so rather the topmost layer hence global-l0. But I understand that in a way you can also think about it as a core layer, because lower layers check back with the topmost layer to get aligned. Anyway for me it's rather the topmost layer and naming it `core` sounds misleading for me. We don't have to change it if we don't agree, especially that it doesn't change any functionality and it's just a name.

Reply:
I agree that `core` sounds misleading. I would go either with `dag-l0` or `global-l0` (preferably the first one)

Reply:
but it's not only a `dag-l0` @tbekas, because `dag-l0` is just a part of this module.